### PR TITLE
roachtest: run multiple backup types in backup/mixed-version

### PR DESCRIPTION
**Pull request by renatolabs**
**roachtest: run multiple backup types in backup/mixed-version**

This adds support to database and full cluster backups in the
`backup/mixed-version` roachtest. When a cluster is taken in
mixed-version state, a randomly chosen backup type is performed. Since
cluster backups also backup some system tables, this commit also adds
logic related to verifying that the contents of system tables are
properly preserved when these cluster backups are restored.

**roachtest: run tpcc workload in backup/mixed-versions**

This commit adds a second workload to the `backup/mixed-versions`
test, namely `tpcc` with 100 warehouses. Previously, we would only run
the `bank` workload; by running TPCC, we verify backups of databases
that includes multiple tables, and the backup of tables that includes
foreign key constraints. It should also put the cluster under more
load than just `bank`, potentially exposing interesting edge cases
that wouldn't come up before.

Epic: CRDB-19321

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
This is a stacked PR, please review them in order (but feel free to review the second one before the first one is merged):

* #100754 
* #100755 **(this PR)**

To be reviewed here: **2nd and 3rd commits**.

**Review by herkolategan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTsS5ThDAUg1GyvjJRX:-NTsS5ThDAUg1GyvjJRY:b-6z2nmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L83)):*
> ```Go
> 	// are included as part of a full cluster backup. It should include
> 	// every table that opts-in to cluster backup (see `system_schema.go`).
> 	// It should be updated as system tables are added or removed from
> ```

Nit: I'm always a bit dubious of scenarios where somebody needs to remember to update something somewhere. I did note this comes from a `ccl` package though - `GetSystemTablesToIncludeInClusterBackup`- which definitely makes it harder to get to. I'm curious as I noticed `cdc` "breaks the rule" by importing from `ccl`. Would this also make the exceptions list?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTsxHeE3YlIbPX-XGxb:-NTsxHeE3YlIbPX-XGxc:bdilvfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L227)):*
> ```Go
> 	tableContents interface {
> 		// Load computes the contents of a table with the given name
> 		// (passed as argument) and stores it representation internally.
> ```

Nit: 'its'
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt4XVEBbYLAWvRyFJx:-NTt4XVEBbYLAWvRyFJy:bum2dmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L293)):*
> ```Go
> 		nonce   string
> 
> 		// tables is the list of tables that were are going to verify once
> ```

Nit: we are
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt5JW_7JsEIJwgaPJO:-NTt5JW_7JsEIJwgaPJP:b-babti6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L330)):*
> ```Go
> //
> //	tableNamesWithDB("db1", []string{"foo.bar, baz"})
> //	=> []string{"db1.foo", "db1.baz"}
> ```

The code doesn't seem to match the example here.

**Review by srosenberg — COMMENTED** [review]
> Reviewed 9 of 9 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @renatolabs)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go` line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTukrsY7-912CszFYXu:-NTukrsY7-912CszFYXv:b4ltw8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/233b035f7baf1d8e4341db0c197ea0c048bbd17f/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go#L229)):*
> ```Go
> 		// cancel the underlying context.
> 		//
> 		// There should be one channel in `bgChans` for each hook in
> ```

Maybe make this stronger? E.g., "Invariant: there is exactly one channel in `bgChans` for each hook in `Test.hooks.background`."
___
*[`pkg/cmd/roachtest/tests/backup.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTumt8D1Odss3EAbFIZ:-NTumt8D1Odss3EAbFI_:bwuv9up) ([raw file](https://github.com/cockroachdb/cockroach/blob/233b035f7baf1d8e4341db0c197ea0c048bbd17f/pkg/cmd/roachtest/tests/backup.go#L166)):*
> ```Go
> 	for j, ip := range ips {
> 		if err := retry.WithMaxAttempts(ctx, retryConfig, 10, func() error {
> 			addr := net.JoinHostPort(ip, fmt.Sprintf("%d", port))
> ```

Nice! 
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTun_9sEkxt7oR0x9gS:-NTun_9sEkxt7oR0x9gT:b-u6ycz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L149)):*
> ```Go
> // tables, we need to quote the column names to avoid SQL syntax
> // errors.
> func quoteColumnsForSelect(columns []string) string {
> ```

Seems like a reusable (i.e., common) helper func.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTup62x6NthiMCx4CMV:-NTup62x6NthiMCx4CMW:bqpwenp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L450)):*
> ```Go
> 
> // Load computes the fingerprints for the underlying table and stores
> // the contents in the `fingeprrints` field.
> ```

`fingerprrints` -> `fingerprints`

Why is this func called `Load`? It's not really loading data. Maybe rename to `ComputeFingerprints`?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 512 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTupzsd6ukf4-zAW12c:-NTupzsd6ukf4-zAW12d:b-njihip) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L512)):*
> ```Go
> // in this system table row, in which case the error should be
> // returned to the caller.
> func (sr *systemTableRow) skip() bool {
> ```

`skipColumn`?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 773 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTurj7J1hfS3WQPYCYk:-NTurj7KCv3VvceHiCeK:bwzm5ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L773)):*
> ```Go
> 	}
> 
> 	sc.rows = rowSet
> ```

Curious how large of a memory footprint do we expect?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 1101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTusC0-3rxlnqXE5T4M:-NTusC0-3rxlnqXE5T4N:b-jr04r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L1101)):*
> ```Go
> 			if strings.HasPrefix(table, "system.") {
> 				node := h.RandomNode(rng, mvb.roachNodes)
> 				contents, err = newSystemTableContents(ctx, mvb.cluster, node, db, table, timestamp)
> ```

Remind me why we can't fingerprint system tables?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 1439 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTuqHwv76zf8qaqdxqO:-NTuqHwv76zf8qaqdxqP:b5oeh9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L1439)):*
> ```Go
> 
> // resetCluster wipes the entire cluster and starts it again with the
> // current latest binary. This is done prior we attempt restoring a
> ```

`prior` -> `before`
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt5JW_7JsEIJwgaPJO:-NTuoRv88Ucf7mvxPYpQ:b1hog5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L330)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

The code doesn't seem to match the example here.
</blockquote></details>

It does, but it's rather strange. We drop `foo` on the floor, which is the strange part, i.e., counter-intuitive.

**Comment by renatolabs** [comment]
Rebased:

* ~#100754~ (merged!)
* #100755 **(this PR)**
* #101730 

To be reviewed here: **the entire diff**.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go` line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTukrsY7-912CszFYXu:-NTxsLnQ9NpCkWowew40:b4ltjb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/233b035f7baf1d8e4341db0c197ea0c048bbd17f/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go#L229)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Maybe make this stronger? E.g., "Invariant: there is exactly one channel in `bgChans` for each hook in `Test.hooks.background`."
</blockquote></details>

I had slightly updated this comment (Herko had a similar suggestion). Updated to include the "Invariant" part too.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTun_9sEkxt7oR0x9gS:-NTxuQnA2V_ny3Pe3Phd:boa3l8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L149)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Seems like a reusable (i.e., common) helper func.
</blockquote></details>

Yes, which is why this function is not tied to any structs used by this test. Since all roachtests share the same package at the moment, they can use this function. Or did you mean I should move the function to an `util.go` file to make its reusability more explicit?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTup62x6NthiMCx4CMV:-NTxuy2W2swG1evVmzks:b-wjc435) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L450)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

`fingerprrints` -> `fingerprints`

Why is this func called `Load`? It's not really loading data. Maybe rename to `ComputeFingerprints`?
</blockquote></details>

Typo fixed.

As for the `Load` function: this is implementing an interface. I chose `Load` in the interface to be the more generic term to describe what happens across user and system tables. IMO, a valid interpretation of this function is that we are loading their contents in different ways: for user tables, the representation of the loaded data is a series of fingerprints. For system tables, we do load the entire contents of the table (as JSON encoded arrays -- implementation detail). I tried to convey that in the comments, but let me know if there's more I can do to make the point clearer.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 512 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTupzsd6ukf4-zAW12c:-NTxw4ZA5m2X29U7TPJp:b-hk5jvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L512)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

`skipColumn`?
</blockquote></details>

While currently this function is only called in the context of `processColumn`, it doesn't need to be so. `skip` indicates that any operation on `systemTableRow` should be skipped (if there was a function that processed entire rows at a time, they should also be skipped if `skip()` returns true.)
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 773 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTurj7J1hfS3WQPYCYk:-NTxwTKv2FDuONp7RON3:bsex5y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L773)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Curious how large of a memory footprint do we expect?
</blockquote></details>

Not much at all; a lot of these system tables are either empty or have very few rows in them (order of couple dozen, max).
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 1101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTusC0-3rxlnqXE5T4M:-NTxx9Le2RWfPDyTv6CU:b-fi8xhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L1101)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Remind me why we can't fingerprint system tables?
</blockquote></details>

Because the test is running an upgrade, and then all bets are off. System tables could have new columns, new rows, etc... Even if there wasn't an upgrade going on, restore logic for system tables can sometimes be quite custom, including "rekeying" of certain keys, which make direct comparison as provided by fingerprint matching not possible. Not to mention impossible to avoid (at roachtest level) races for scheduled jobs. There's probably more reasons I'm not remembering.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 1439 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTuqHwv76zf8qaqdxqO:-NTxyIuI-urPFw-nTWQG:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4e4920ab1eee5863cd632078d1c144ee97cc2d3/pkg/cmd/roachtest/tests/mixed_version_backup.go#L1439)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

`prior` -> `before`
</blockquote></details>

Fixed, thanks.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTsS5ThDAUg1GyvjJRX:-NTxsgPk9bYDlB2QrcKK:b-1m5bz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L83)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: I'm always a bit dubious of scenarios where somebody needs to remember to update something somewhere. I did note this comes from a `ccl` package though - `GetSystemTablesToIncludeInClusterBackup`- which definitely makes it harder to get to. I'm curious as I noticed `cdc` "breaks the rule" by importing from `ccl`. Would this also make the exceptions list?
</blockquote></details>

I kind of like the explicitness of having the entire list spelled out in the test.

To verify the contents of system tables, especially in the context of an upgrade, the test is required to be aware of certain special behavior associated with the restoring of these tables (such as ignoring the`version` row in `system.settings`, for example). In addition, we also need to ignore the `system.zones` table at the moment (see associated comment in this file). So even if we imported the list of names from the backup package, we'd still need some manually written logic to deal with these special cases. And then at that point, it's arguable whether the benefit is still there. If, at some point, we stop backing up the `system.settings` table, as an arbitrary example, the test would continue to pass but we now have dead code in the test dealing with special cases in the restore that no longer apply.

If there are changes in the system tables we backup, I'd see a failure in this test as a "feature"; we would look at the failure and acknowledge it's expected. The fix would likely be straightforward.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTsxHeE3YlIbPX-XGxb:-NTxunN7C0phvelPj89y:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L227)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: 'its'
</blockquote></details>

Fixed, thanks.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt4XVEBbYLAWvRyFJx:-NTxuoq23bYXnpxSe9nS:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L293)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: we are
</blockquote></details>

Fixed, thanks.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt5JW_7JsEIJwgaPJO:-NTxuqHN-BVawCJH20Pv:b7g5srs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L330)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

It does, but it's rather strange. We drop `foo` on the floor, which is the strange part, i.e., counter-intuitive.
</blockquote></details>

Yeah, it does match the example. I updated the documentation for this function, let me know if it's clearer now.

**Review by herkolategan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt5JW_7JsEIJwgaPJO:-NTwkxcE3zLtCVgaywJu:b-i3byf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L330)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Yeah, it does match the example. I updated the documentation for this function, let me know if it's clearer now.
</blockquote></details>

Thanks, but what I meant, as with the original example, is that the input `tableNamesWithDB("restored_db", []string{"bank.bank, stock"})` seems to suggest that it can support a comma separated string as well as a slice.
In other words shouldn't it be `tableNamesWithDB("restored_db", []string{"bank.bank", "stock"})`? In the current state the example would output: `"restored_db.bank, stock"`

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NTt5JW_7JsEIJwgaPJO:-NU2-xzm9SRUhoRqOu1J:b-y5xyvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dd1c3c09fde916183ba019e108dc66e636cf534/pkg/cmd/roachtest/tests/mixed_version_backup.go#L330)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Thanks, but what I meant, as with the original example, is that the input `tableNamesWithDB("restored_db", []string{"bank.bank, stock"})` seems to suggest that it can support a comma separated string as well as a slice.
In other words shouldn't it be `tableNamesWithDB("restored_db", []string{"bank.bank", "stock"})`? In the current state the example would output: `"restored_db.bank, stock"`
</blockquote></details>

Oh, I see, I didn't notice that! I often make that typo, thanks for catching. Fixed.

**Review by renatolabs — COMMENTED** [review]
Temporary commit dropped, this is back to what should be merged.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-:-NU205j60XxY_UNx-727:b-tsx0y3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

**Review by herkolategan — APPROVED** [review]

**Review by smg260 — COMMENTED** [review]
> Reviewed 1 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NU2jxcv0SwGDv0ZR9Lu:-NU2jxcv0SwGDv0ZR9Lv:b-r0aj55) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f274c64a30bcda306d9ec3dd804e450cced7710/pkg/cmd/roachtest/tests/mixed_version_backup.go#L220)):*
> ```Go
> 	// tableContents is an interface to be implemented by different
> 	// methods of storing and comparing the contents of a table. A lot
> 	// of backup-related tests use the `EXPERIMENTAL_FINGEPRINTS`
> ```

typo: `EXPERIMENTAL_FINGE[R]PRINTS`
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NU2kS0v3WpEdC_p1DZN:-NU2kS0v3WpEdC_p1DZO:b-68owy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f274c64a30bcda306d9ec3dd804e450cced7710/pkg/cmd/roachtest/tests/mixed_version_backup.go#L239)):*
> ```Go
> 
> 	// fingerprintContents implements the `tableContents` interface
> 	// using the `EXPERIMENTAL_FINGEPRINTS` functionality. This should
> ```

typo: as above

**Comment by smg260** [comment]
Is there a better way to organise the code in `mixed_version_backup.go`, which is getting really long? Maybe it all makes sense together. Perhaps separate helpers from the meat of the test?

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NU2jxcv0SwGDv0ZR9Lu:-NU3B6lh8fDH8ZQ8PE6o:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f274c64a30bcda306d9ec3dd804e450cced7710/pkg/cmd/roachtest/tests/mixed_version_backup.go#L220)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

typo: `EXPERIMENTAL_FINGE[R]PRINTS`
</blockquote></details>

Fixed, thanks.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-NU2kS0v3WpEdC_p1DZN:-NU3BAX48UFnUP2OsQEx:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f274c64a30bcda306d9ec3dd804e450cced7710/pkg/cmd/roachtest/tests/mixed_version_backup.go#L239)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

typo: as above
</blockquote></details>

Fixed, thanks.

**Review by renatolabs — COMMENTED** [review]
Indeed, it is becoming quite long. I'm not sure how much can be moved to a different file, especially given the current roachtest organization where every test is under the same package (does it add more confusion if we split a test in multiple files?).

I've talked to DR about making some of the functionality in this test more generally applicable to non-mixed-version backup/restore tests. I think it makes sense to defer any moving of code to different files at that point.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100755#-:-NU3BGPj6A9KRFwUBCVx:bhy6kxd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

**Comment by renatolabs** [comment]
I expect this to flake due to connection warmup hanging (to be fixed by #101786). I'm going to merge it anyway so that we get some runs in; flakes should be pretty easy to identify.

TFTRs!

bors r=herkolategan
